apache2/msc_multipart.c

```diff
@@ -198,23 +198,44 @@ static int multipart_parse_content_disposition(modsec_rec *msr, char *c_d_value)
                     log_escape_nq(msr->mp, value));
             }
         }
-        else
-        if (strcmp(name, "filename") == 0) {
+        else if ((strcmp(name, "filename") == 0) || (strcmp(name, "filename*") == 0))
+        {
 
-            validate_quotes(msr, value);
+            char *decoded_filename = NULL;
 
-            msr->multipart_filename = apr_pstrdup(msr->mp, value);
+            if (strcmp(name, "filename*") == 0)
+            {
+                decoded_filename = rfc5987_decode(msr->mp, value);
+                if (!decoded_filename)
+                {
+                    msr_log(msr, 4, "Multipart: Could not decode extended filename parameter in RFC 5987 format: %s",
+                            log_escape_nq(msr->mp, value));
+                    return -16;
+                }
+                msr->multipart_filename = decoded_filename;
```

Comment:
Looks like they are calling apr_pstrdup in the original code. Maybe that allocates a new buffer for this string, in case "value" goes out of scope and gets overwritten?

Reply:
yeah apr_pstrdup allocates a new buffer from the given pool and copies over the string. We already created the buffer and copied over the string `rfc5987_decode` so here we are just copying over the pointer.